### PR TITLE
Create a general error handler which handles multiple errors - Nia 1435

### LINKED ISSUE #1435: Create a general error handler which handles multiple errors
Introduce a general error handler which displays errors using a SnackBar one after each other ([following the Material guidelines](https://m3.material.io/components/snackbar/guidelines#84cd5bb6-354f-4fd0-bdd9-daf9684ba4bf)).

Context: https://github.com/android/nowinandroid/pull/1090

### PR DESCRIPTION
**Pull request by tjmtic**
Fixes #1435  

Simple and Generic Error Handler, with a few extensible error types and an OFFLINE error (per reference to #350 ). Not exactly sure if there was something more specific to be implementing or errors to cover, so will be curious if there is a different approach here.

Included in core/ui due to the dependency reference to SnackbarHostState.

The errorHandler can be passed down through composables from NiaApp, the same way as onShowSnackbar, and then called when necessary by the ui. I am not confident in the instantiation in NiaApp, but am not sure it would be possible before then since the handler requires the snackbarHostState in this implementation.

### CODE REVIEW

File: core/ui/src/main/kotlin/com/google/samples/apps/nowinandroid/core/ui/SnackbarErrorHandler.kt [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        // Log the error or show a generic error message
+        when (error) {
+            is Specific -> {
+                snackbarHostState.showSnackbar(error.data.toString())
+            }
+            is Exception -> {
+                snackbarHostState.showSnackbar("Exception: ${error.exception.message}")
+            }
+            is Offline -> {
+                snackbarHostState.showSnackbar("No internet connection")
```
Comment (lihenggui) [inline core/ui/src/main/kotlin/com/google/samples/apps/nowinandroid/core/ui/SnackbarErrorHandler.kt:37]: Strings should be in the `strings.xml` file

**Comment by anhtuannd** [comment]
But still only one error can be displayed at a time. When the device is offline, will other errors be shown and replaced offline error?

File: app/src/main/kotlin/com/google/samples/apps/nowinandroid/navigation/NiaNavHost.kt [hunk 2]
```diff
@@ -39,6 +40,7 @@ import com.google.samples.apps.nowinandroid.ui.interests2pane.interestsListDetai
 fun NiaNavHost(
     appState: NiaAppState,
     onShowSnackbar: suspend (String, String?) -> Boolean,
+    snackbarErrorHandler: SnackbarErrorHandler,
```
Comment (anhtuannd) [inline app/src/main/kotlin/com/google/samples/apps/nowinandroid/navigation/NiaNavHost.kt:43]: This is unused.

File: core/ui/src/main/kotlin/com/google/samples/apps/nowinandroid/core/ui/SnackbarErrorHandler.kt [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+import androidx.compose.material3.SnackbarHostState
+import com.google.samples.apps.nowinandroid.core.ui.SnackbarError.Default
+import com.google.samples.apps.nowinandroid.core.ui.SnackbarError.Exception
+import com.google.samples.apps.nowinandroid.core.ui.SnackbarError.Offline
+import com.google.samples.apps.nowinandroid.core.ui.SnackbarError.Specific
+
+class SnackbarErrorHandler(private val snackbarHostState: SnackbarHostState) {
+
+    suspend fun handleError(error: SnackbarError<*>) {
```
Comment (anhtuannd) [inline core/ui/src/main/kotlin/com/google/samples/apps/nowinandroid/core/ui/SnackbarErrorHandler.kt:27]: It can be top level function

**Review by anhtuannd — COMMENTED** [review]
I think the PR is not completed yet.

File: core/ui/src/main/kotlin/com/google/samples/apps/nowinandroid/core/ui/SnackbarErrorHandler.kt (hunk 1 above)
Comment (tjmtic) [reply core/ui/src/main/kotlin/com/google/samples/apps/nowinandroid/core/ui/SnackbarErrorHandler.kt]: Yes, moved these, thank you.

File: app/src/main/kotlin/com/google/samples/apps/nowinandroid/navigation/NiaNavHost.kt (hunk 2 above)
Comment (tjmtic) [reply app/src/main/kotlin/com/google/samples/apps/nowinandroid/navigation/NiaNavHost.kt]: Yes, removed unused, thank you.

File: core/ui/src/main/kotlin/com/google/samples/apps/nowinandroid/core/ui/SnackbarErrorHandler.kt (hunk 3 above)
Comment (tjmtic) [reply core/ui/src/main/kotlin/com/google/samples/apps/nowinandroid/core/ui/SnackbarErrorHandler.kt]: Thank you, updated this implementation.

**Comment by tjmtic** [comment]
Going to close this PR. 

After looking more at the [ comment in the linked issue ](https://github.com/android/nowinandroid/pull/1090#discussion_r1433979886)
I think I see what the intended approach was and will update with a new PR.

**Pull request closed without merging**